gapic/samplegen/samplegen.py

```diff
@@ -0,0 +1,288 @@
+# Copyright (C) 2019  Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import itertools
+import keyword
+import re
+import gapic.samplegen.utils
+
+from collections import defaultdict
+
+# Outstanding issues:
+# * Neither license nor copyright are defined
+# * In real sample configs, many variables are
+#   defined with an _implicit_ $resp variable.
+
+MIN_SCHEMA_VERSION = (1, 2, 0)
+
+VALID_CONFIG_TYPE = 'com.google.api.codegen.SampleConfigProto'
+
+# There are a couple of other names that should be reserved
+# for sample variable assignments, e.g. 'client', but there are diminishing returns:
+# the sample config author isn't a pathological adversary, they're a normal person
+# who may make mistakes occasionally.
+RESERVED_WORDS = frozenset(itertools.chain(keyword.kwlist, dir(__builtins__)))
+
+TEMPLATE_NAME = 'samplegen_template.j2'
+
+
+class ReservedVariableName(Exception):
+    pass
+
+
+class SchemaVersion(Exception):
+    pass
+
+
+class RpcMethodNotFound(Exception):
+    pass
+
+
+class InvalidConfigType(Exception):
+    pass
+
+
+class InvalidStatement(Exception):
+    pass
+
+
+class BadLoop(Exception):
+    pass
+
+
+class MismatchedPrintStatement(Exception):
+    pass
+
+
+class UndefinedVariableReference(Exception):
+    pass
+
+
+class BadAssignment(Exception):
+    pass
+
+
+class InconsistentRequestName(Exception):
+    pass
+
+
+class InvalidRequestSetup(Exception):
+    pass
+
+
+class DuplicateInputParameter(Exception):
+    pass
+
+
+def validate_response(response):
+    coll_kword = "collection"
+    var_kword = "variable"
+    map_kword = "map"
+    key_kword = "key"
+    val_kword = "value"
+    body_kword = "body"
+    # Do NOT need checks for valid lexical scoping
+    # because of python variable definition rules.
+    # E.g. the following is valid python:
+    #
+    # for i in range(10):
+    #   squid = "Squid {}".format(i)
+    #
+    # print("Last squid is {}".format(squid))
+
+    def handle_define(body):
+        m = re.match(r"^([^=]+)=([^=]+)$", body)
+        if not m:
+            raise BadAssignment("Bad assignment statement: {}", body)
+
+        var, rval = m.groups()
+        if var in RESERVED_WORDS:
+            raise ReservedVariableName(
+                "Tried to define a variable with reserved name: {}".format(var))
```

Comment:
Agreed that @vchudnov-g and @hzyi-google should weigh in.

I would argue that erroring out is the better thing to do. It's cleaner, easier, and makes a more explicable mapping between the sample config and generated sample. Also, while the set of reserved keywords in all languages is technically infinite, there's going to be a lot of overlap, and the languages of importance are going to be well defined and closed for the near future.
The sample config isn't an immutable artifact in the same way the API proto is: it can be modified relatively painlessly by the author to not use keywords in whatever languages they care about.

Reply:
I think erroring out is the right answer here.

In short:
- erroring out forces the sample in each language to be idiomatic, even if the internal variable names/CLI args don't quite match across languages
- silently fixing this instead of erroring out unnecessarily couples sample-tester to the disambiguation algorithm implemented in the generator(s)


To elaborate:

Let's say that the config has a conflict just for language "Lore". There are two scenarios of interest here: CONCURRENT, generating Lore samples at the same time as other languages, and SEQUENTIAL, generating Lore samples after other languages have had their samples generated successfully.

In CONCURRENT:

* If we error, then sample owners can either ignore Lore altogether for now, in which case they jump to SEQUENTIAL, or fix the sample config and test config (the latter if the conflict occurs in a CLI argument). By fixing the config:
  - we support idiomaticity: the identifier in Lore is chosen in such a way that it does not resemble the keyword
  - all the sample variables are parallel across languages, which is desirable though not required (highly desirable for CLI args)

* If we silently fix this:
  - the identifier in Lore resembles a keyword, which is not at all idiomatic,
  - sample authors are not forced to be made aware that the identifier they chose was a keyword, and that the one actually used is very close to a keyword
  - all the sample variables remain roughly parallel across languages, save for the disambiguation

In SEQUENTIAL:
* If we error, then the sample config owners are forced to fix the config. By fixing the config:
  - we support idiomaticity: the identifier in Lore is chosen in such a way that it does not resemble the keyword
  - if the owners regenerate the samples for all languages: we are back to consequences of CONCURRENT
  - if the owners instead regenerate the samples for Lore only:
     - the variable name used internally in Lore is different than the variable name used in other languages. This is less pretty.
     - if the conflict occurred in a CLI arg, sample invocation changes across languages. Not ideal, but IMO not a deal breaker, since typically folks can get help on usage.
     - the previously generated and tested samples continue to work
     - if the tests needed to be updated because the conflict was in a CLI arg, then the old samples will fail the new tests due to the new CLI arg name. I would argue that is not surprising
  
* If we silently fix this:
  - the identifier in Lore resembles a keyword, which is not at all idiomatic,
  - sample authors are not forced to be made aware that the identifier they chose was a keyword, and that the one used is very close to a keyword
  - all the sample variables remain roughly parallel across languages, save for the disambiguation
  - in the case of the conflict occurring in a CLI argument, sample-tester needs to somehow know that in a particular language, a particular argument name needs to be modified in a certain way in order to match the auto-fixed CLI arg name. This is redundant and very brittle, since it forces the resolution algorithm to match between sample-tester and the generators.

Reply:
Going for erroring out.